compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -208,20 +208,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         // without the final expr (e.g. `try { return; }`). We don't want to generate an
         // unreachable_code lint for it since warnings for autogenerated code are confusing.
         let is_try_block_generated_unit_expr = match expr.kind {
-            ExprKind::Call(_, args) if expr.span.is_desugaring(DesugaringKind::TryBlock) => {
-                args.len() == 1 && args[0].span.is_desugaring(DesugaringKind::TryBlock)
+            ExprKind::Call(_, [arg]) => {
+                expr.span.is_desugaring(DesugaringKind::TryBlock)
+                    && arg.span.is_desugaring(DesugaringKind::TryBlock)
             }
-
             _ => false,
         };
 
-        // Warn for expressions after diverging siblings.
-        if !is_try_block_generated_unit_expr {
-            self.warn_if_unreachable(expr.hir_id, expr.span, "expression");
-        }
+        let old_diverges = if self.is_whole_body_and_arguments_diverge.replace(false) {
```

Comment:
Wait, what exactly is this doing? Pls leave a more detailed comment.

Reply:
Yeah sorry I'm finding this hard to make clear. How's the new version?